### ISSUE TITLE
Possible regression in foreign trait implementations from Swift

### ISSUE THREAD

**Issue by ianthetechie** [mozilla/uniffi-rs#1846]
I was previously using a prerelease commit (b369e7c15b1b7ebca34de9028209db11b7ff353d) to get some fixes for proc macros. It seems that sometime between this commit and release, code generation is broken so that I can't provide a foreign implementation of a trait to a constructor anymore. I'll include some abbreviated code extracts below.

Trait implementations look like this:

```rust
#[uniffi::export]
pub trait RouteRequestGenerator: Send + Sync {
    fn generate_request(
        &self,
        user_location: UserLocation,
        waypoints: Vec<GeographicCoordinates>,
    ) -> Result<RouteRequest, RoutingRequestGenerationError>;
}
```

Object definition that stores one of these dyn trait references:

```rust
#[derive(uniffi::Object)]
pub struct RouteAdapter {
    request_generator: Arc<dyn RouteRequestGenerator>,
    response_parser: Arc<dyn RouteResponseParser>,
}

#[uniffi::export]
impl RouteAdapter {
    #[uniffi::constructor]
    pub fn new(
        request_generator: Arc<dyn RouteRequestGenerator>,
        response_parser: Arc<dyn RouteResponseParser>,
    ) -> Arc<Self> {
        Arc::new(Self {
            request_generator,
            response_parser,
        })
    }
    // ...
}
```

I would expect to be able to create a `RouteRequestGenerator` trait implementation in Swift by conforming to the `RouteRequestGeneratorProtocol` in v0.25.1, and be able to pass this as a constructor argument. In a previous commit, the generated bindings used the names `RouteRequestGenerator` for the protocol and `RouteRequestGeneratorImpl` for the implementation. Here is the Xcode diff of the generated bindings for our project.

<img width="633" alt="image" src="https://github.com/mozilla/uniffi-rs/assets/2140440/466542aa-8230-4071-809b-4c4a925944e3">

However, the `RouteAdapter` code gen didn't change in the intervening commits, so the constructor expects a concrete class rather than the protocol type (the name didn't change, and it used to be that the protocol was named this way). The implementation is (correctly) not an `open class` so it's not currently possible, as far as I can tell, to provide a Swift implementation.

<img width="620" alt="image" src="https://github.com/mozilla/uniffi-rs/assets/2140440/96980854-91c7-4bef-aa39-1cc0a53f8b83">

What I would expect to work code-wise:

```swift
private class MockRouteRequestGenerator: RouteRequestGeneratorProtocol {
    func generateRequest(userLocation: UniFFI.UserLocation, waypoints: [UniFFI.GeographicCoordinates]) throws -> UniFFI.RouteRequest {
        return UniFFI.RouteRequest.httpPost(url: backendUrl.absoluteString, headers: [:], body: Data())
    }
}

// Error: Cannot convert value of type 'MockRouteRequestGenerator' to expected argument type 'RouteRequestGenerator'
let routeAdapter = RouteAdapter(requestGenerator: MockRouteRequestGenerator(), responseParser: MockRouteResponseParser(routes: []))
```

**Comment by ianthetechie** [comment]
I just tried the same thing in Kotlin and hit a similar result. So either it's consistent, or (probably equally likely) I just missed something / the docs need some updates on how to do this sort of thing.

```kotlin
class MockRouteRequestGenerator: RouteRequestGeneratorInterface {
    override fun generateRequest(
        userLocation: UserLocation,
        waypoints: List<GeographicCoordinates>
    ): RouteRequest = RouteRequest.HttpPost(valhallaEndpointUrl, mapOf(), byteArrayOf())

}

class MockRouteResponseParser(private val routes: List<Route>) : RouteResponseParserInterface {
    override fun parseResponse(response: ByteArray): List<Route> = routes
}

// Type mismatch: requires the non-interface variants.
val core = FerrostarCore(
    routeAdapter = RouteAdapter(requestGenerator = MockRouteRequestGenerator(), responseParser = MockRouteResponseParser()),
    locationProvider = SimulatedLocationProvider(),
    httpClient = OkHttpClient.Builder().addInterceptor(interceptor).build()
)
```

`RouteRequestGenerator`, the type it wants, is a `sealed class` in Kotlin.

Looking over the examples, it seems like there are already test cases covering this fairly closely. FancyTelephone, which uses SimCard: https://github.com/mozilla/uniffi-rs/blob/c6ddceea437423b39eee0e706743ced5237b226f/examples/callbacks/src/lib.rs#L72

There doesn't appear to be a test case involving a trait that's defined via `#[uniffi::export]` so maybe that's the difference? It looks like your unit test subclasses `SimCard` directly rather than the protocol, which is disallowed in both Swift and Kotlin in the case of UniFFI `0.25.1` + the proc macro on the trait (definition in the first comment above): https://github.com/mozilla/uniffi-rs/blob/c6ddceea437423b39eee0e706743ced5237b226f/examples/callbacks/tests/bindings/test_callbacks.swift#L41

That it works over there suggests either 1) a subtle difference in code paths, or 2) that I am missing some detail in the annotations or types.

**Comment by ianthetechie** [comment]
It looks like this is fixed again in main as of `41e3f953ed6e3d8e10cde262dd269b683ac49f0e`; guessing some changes got reverted.
